### ISSUE TITLE
Is it possible to use domain 0 to discover participants of other domains？

### ISSUE THREAD

**Issue by luxingxiao** [eclipse-cyclonedds/cyclonedds#1566]
Using the ls command of python can only see the participants of the same domain at present. If I want to automatically discover the participants of all domains, can I discover them through domain 0?

**Comment by eboasson** [comment]
(Copy-paste from my response on Discord, for those who happen to be searching for something similar on GitHub)

I don't think so: a normal domain 0 participant uses port numbers different from a domain X≠0 participant, so it won't even see the traffic. We also include the domain id in the discovery messages so that it won't talk even when the port numbers are the same.

So, even though you can override the port number mapping, it still wouldn't work. (There is also a configuration option to change the advertised domain id to, for example, 0, but then you're effectively domain 0.)

What could possibly be an option is to create two participants, one in domain 0 that you only use to publish/subscribe information on participants in other domains — that'd be pretty straightforward with a transient-local topic. I don't think that's what you're looking for though ...

You could do a hack, if you really want to. Run all domains on the same port number and hack the domain id check so the domain 0 participant ignores the advertised domain id. That's a really small hack, see https://github.com/eclipse-cyclonedds/cyclonedds/blob/de27a112cf673d9479d096ce4a9ac17f9d280434/src/core/ddsi/src/ddsi_discovery.c#L737

Probably not going to make it into the code base, but if the argument is strong enough, then who knows 🙂

**Comment by i-and** [comment]
For example, for all domains (except 0), you can additionally send a special not RTPS message to port 7400, which will contain the domain number. When receiving this message (in the context of a participant with domain number 0), the only thing the stack does is callback the user with the discovered domain number. Then, if necessary, the user creates a Participant within discovered domain number using the normal Cyclone DDS API.

**Comment by eboasson** [comment]
> For example, for all domains (except 0), you can additionally send a special not RTPS message to port 7400, which will contain the domain number. When receiving this message (in the context of a participant with domain number 0), the only thing the stack does is callback the user with the discovered domain number. Then, if necessary, the user creates a Participant within discovered domain using the normal Cyclone DDS API.

Something of the sort could indeed be done. It wouldn't even have to be a non-RTPS message, one could just use a vendor-specific submessage (or even the full SPDP message, but that creates rather more overhead). I suspect it would be a good application for a [gossiping protocol](https://research.utwente.nl/en/publications/gossiping-in-distributed-systems) to do the discovery reasonably quickly with very low cost. There are so many fun things to try out and so little time ...

Another possibility along the same lines might be to create a participant in a separate domain that has no endpoints other than SPDP itself. The protocol and the implementation can handle it in principle [^1], so it really is but a small change in the API (or perhaps the configuration). Then you get the same with minimal effort.

[^1]: If the `flags` parameter of [`ddsi_new_participant`](https://github.com/eclipse-cyclonedds/cyclonedds/blob/09a02f5b2b9ba856d7d4c616510252cf9e508ec2/src/core/ddsi/src/ddsi_participant.c#L1017) has `RTPS_PF_NO_BUILTIN_READERS` and `RTPS_PF_NO_BUILTIN_WRITERS` set it won't create any discovery readers/writers. It isn't exactly what you need but it does come close.

**Comment by i-and** [comment]
Indeed, on 239.255.0.1:7400, each regular Participant could additionally send a message similar to DATA(p), but with an edit in the header, which leads to the message ceasing to be accepted by ordinary SPDP readers (this I understood by "special not RTPS message"). To save traffic, you can send it with a longer period and make it shorter and leave only the DOMAIN_NUMBER parameter there.

In general, it seems to me that this domain detection function in the system is very much in demand, especially in DDS network monitoring tools. Without this, the user either needs to know all the domain numbers in the system (which is not always possible) or sort through the entire range of ports, which is very time-consuming and unreliable.

**Comment by Dominik-38** [comment]
Hello, I just wanted to ask if this issue is still ongoing and will be processed in the near future?
I am also interested in something like a domain detection function, which lists all active domains for analysis.
